# **Ecommerce Data set**
**Complete Exploratory Data Analysis**

## About Dataset

>- **`Description`**\
> The Data Set was downloaded from Kaggle, from the following [link](https://www.kaggle.com/datasets/zusmani/pakistans-largest-ecommerce-dataset/data)


- `Context`
  
  This is the largest retail e-commerce orders dataset from Pakistan. It contains half a million transaction records from March 2016 to August 2018. The data was collected from various e-commerce merchants as part of a research study. I am releasing this dataset as a capstone project for my data science course at Alnafi (alnafi.com/zusmani).
There is a dire need for such dataset to learn about Pakistan’s emerging e-commerce potential and I hope this will help many startups in many ways.

- `Content`
  
  Geography: Pakistan

Time period: 03/2016 – 08/2018

Unit of analysis: E-Commerce Order

- `Acknowledgements`
  
  I like to thank all the startups who are trying to make their mark in Pakistan despite the unavailability of research data.

- `Inspiration`

I’d like to call the attention of my fellow Kagglers to use Machine Learning and Data Sciences to help me explore these ideas:

• What is the best-selling category?
• Visualize payment method and order status frequency
• Find a correlation between payment method and order status
• Find a correlation between order date and item category
• Find any hidden patterns that are counter-intuitive for a layman
• Can we predict number of orders, or item category or number of customers/amount in advance?

## 1. ***Importing Libraries***

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline 
# hide all warnings runtime
import warnings
warnings.filterwarnings('ignore')

# 2. **Data Loading, Exploration, and Cleansing**

 ↪ Load the CSV file into a Pandas DataFrame

 ↪ Investigate and comprehend the data within the dataset using Pandas

 ↪ Address the following data quality concerns:
 
- Handle missing data
- Identify and manage outliers
- Rectify any incorrect or invalid records

In [2]:
df = pd.read_csv("Pakistan Largest Ecommerce Dataset.csv")

- Lets have a look on fist 5 some columns and rows 

In [3]:
df.head(5)

item_id          status created_at  \
0  211131.0        complete   7/1/2016   
1  211133.0        canceled   7/1/2016   
2  211134.0        canceled   7/1/2016   
3  211135.0        complete   7/1/2016   
4  211136.0  order_refunded   7/1/2016   

                                                 sku   price  qty_ordered  \
0                                  kreations_YI 06-L  1950.0          1.0   
1  kcc_Buy 2 Frey Air Freshener & Get 1 Kasual Bo...   240.0          1.0   
2                                 Ego_UP0017-999-MR0  2450.0          1.0   
3                                     kcc_krone deal   360.0          1.0   
4                                        BK7010400AG   555.0          2.0   

   grand_total increment_id    category_name_1 sales_commission_code  ...  \
0       1950.0    100147443    Women's Fashion                    \N  ...   
1        240.0    100147444  Beauty & Grooming                    \N  ...   
2       2450.0    100147445    Women's Fashion                    \N  ...   
3         60.0    100147446  Beauty & Grooming           R-FSD-52352  ...   
4       1110.0    100147447            Soghaat                    \N  ...   

   Month Customer Since     M-Y    FY Customer ID  Unnamed: 21  Unnamed: 22  \
0    7.0         2016-7  7-2016  FY17         1.0          NaN          NaN   
1    7.0         2016-7  7-2016  FY17         2.0          NaN          NaN   
2    7.0         2016-7  7-2016  FY17         3.0          NaN          NaN   
3    7.0         2016-7  7-2016  FY17         4.0          NaN          NaN   
4    7.0         2016-7  7-2016  FY17         5.0          NaN          NaN   

  Unnamed: 23 Unnamed: 24 Unnamed: 25  
0         NaN         NaN         NaN  
1         NaN         NaN         NaN  
2         NaN         NaN         NaN  
3         NaN         NaN         NaN  
4         NaN         NaN         NaN  

[5 rows x 26 columns]

> ***Note*** In order to show the complete output of the rows and columns. We used these commands


In [4]:
pd.set_option('display.max_columns', None, 'display.max_rows', None)

In [5]:
df.head(5)

item_id          status created_at  \
0  211131.0        complete   7/1/2016   
1  211133.0        canceled   7/1/2016   
2  211134.0        canceled   7/1/2016   
3  211135.0        complete   7/1/2016   
4  211136.0  order_refunded   7/1/2016   

                                                 sku   price  qty_ordered  \
0                                  kreations_YI 06-L  1950.0          1.0   
1  kcc_Buy 2 Frey Air Freshener & Get 1 Kasual Bo...   240.0          1.0   
2                                 Ego_UP0017-999-MR0  2450.0          1.0   
3                                     kcc_krone deal   360.0          1.0   
4                                        BK7010400AG   555.0          2.0   

   grand_total increment_id    category_name_1 sales_commission_code  \
0       1950.0    100147443    Women's Fashion                    \N   
1        240.0    100147444  Beauty & Grooming                    \N   
2       2450.0    100147445    Women's Fashion                    \N   
3         60.0    100147446  Beauty & Grooming           R-FSD-52352   
4       1110.0    100147447            Soghaat                    \N   

   discount_amount payment_method Working Date BI Status      MV     Year  \
0              0.0            cod     7/1/2016     #REF!   1,950   2016.0   
1              0.0            cod     7/1/2016     Gross     240   2016.0   
2              0.0            cod     7/1/2016     Gross   2,450   2016.0   
3            300.0            cod     7/1/2016       Net     360   2016.0   
4              0.0            cod     7/1/2016     Valid   1,110   2016.0   

   Month Customer Since     M-Y    FY  Customer ID  Unnamed: 21  Unnamed: 22  \
0    7.0         2016-7  7-2016  FY17          1.0          NaN          NaN   
1    7.0         2016-7  7-2016  FY17          2.0          NaN          NaN   
2    7.0         2016-7  7-2016  FY17          3.0          NaN          NaN   
3    7.0         2016-7  7-2016  FY17          4.0          NaN          NaN   
4    7.0         2016-7  7-2016  FY17          5.0          NaN          NaN   

   Unnamed: 23  Unnamed: 24  Unnamed: 25  
0          NaN          NaN          NaN  
1          NaN          NaN          NaN  
2          NaN          NaN          NaN  
3          NaN          NaN          NaN  
4          NaN          NaN          NaN

- Lets have a look on the names of the columns in the dataset

In [6]:
df.columns

Index(['item_id', 'status', 'created_at', 'sku', 'price', 'qty_ordered',
       'grand_total', 'increment_id', 'category_name_1',
       'sales_commission_code', 'discount_amount', 'payment_method',
       'Working Date', 'BI Status', ' MV ', 'Year', 'Month', 'Customer Since',
       'M-Y', 'FY', 'Customer ID', 'Unnamed: 21', 'Unnamed: 22', 'Unnamed: 23',
       'Unnamed: 24', 'Unnamed: 25'],
      dtype='object')

- Lets check out datatypes, no of columns, rows

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 26 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   item_id                584524 non-null  float64
 1   status                 584509 non-null  object 
 2   created_at             584524 non-null  object 
 3   sku                    584504 non-null  object 
 4   price                  584524 non-null  float64
 5   qty_ordered            584524 non-null  float64
 6   grand_total            584524 non-null  float64
 7   increment_id           584524 non-null  object 
 8   category_name_1        584360 non-null  object 
 9   sales_commission_code  447346 non-null  object 
 10  discount_amount        584524 non-null  float64
 11  payment_method         584524 non-null  object 
 12  Working Date           584524 non-null  object 
 13  BI Status              584524 non-null  object 
 14   MV                    584524 non-

 **Observations**

- No of Columns = 26
- No of Rows    = 1048575 
- So here in the data set some of the columns are unnamed and they are not much important so we can drop these columns.


 Dropping the `Unnamed` columns

In [8]:
# Identify the unnamed columns
column_names = df.columns
unnamed_columns = [col for col in column_names if 'Unnamed' in col]
# Drop the unnamed columns
df = df.drop(unnamed_columns, axis=1)

# Save the updated dataset
df.to_csv('updated_dataset.csv', index=False)

Lets check whether they are dropped or not

In [9]:
df.columns

Index(['item_id', 'status', 'created_at', 'sku', 'price', 'qty_ordered',
       'grand_total', 'increment_id', 'category_name_1',
       'sales_commission_code', 'discount_amount', 'payment_method',
       'Working Date', 'BI Status', ' MV ', 'Year', 'Month', 'Customer Since',
       'M-Y', 'FY', 'Customer ID'],
      dtype='object')

`Note` Here are some full forms of the columns
- sku = stock Keeping Unit
- MV  = Market Value
- M-Y = Month-Year
- FY  = Fiscal Year

Lets have a check on the data types 

In [10]:
df.dtypes

item_id                  float64
status                    object
created_at                object
sku                       object
price                    float64
qty_ordered              float64
grand_total              float64
increment_id              object
category_name_1           object
sales_commission_code     object
discount_amount          float64
payment_method            object
Working Date              object
BI Status                 object
 MV                       object
Year                     float64
Month                    float64
Customer Since            object
M-Y                       object
FY                        object
Customer ID              float64
dtype: object

# 3. Checking Null Values And Duplicates

In [11]:
df.isnull().sum()

item_id                  464051
status                   464066
created_at               464051
sku                      464071
price                    464051
qty_ordered              464051
grand_total              464051
increment_id             464051
category_name_1          464215
sales_commission_code    601229
discount_amount          464051
payment_method           464051
Working Date             464051
BI Status                464051
 MV                      464051
Year                     464051
Month                    464051
Customer Since           464062
M-Y                      464051
FY                       464051
Customer ID              464062
dtype: int64

- Checking duplicates first

In [12]:
df.duplicated().sum()

np.int64(464050)

- A lot of `Duplicates` lets remove them

In [13]:
df = df.drop_duplicates()

In [14]:
df.duplicated().sum()

np.int64(0)

- Now lets check `Null values`

In [15]:
df.isnull().sum()

item_id                       1
status                       16
created_at                    1
sku                          21
price                         1
qty_ordered                   1
grand_total                   1
increment_id                  1
category_name_1             165
sales_commission_code    137179
discount_amount               1
payment_method                1
Working Date                  1
BI Status                     1
 MV                           1
Year                          1
Month                         1
Customer Since               12
M-Y                           1
FY                            1
Customer ID                  12
dtype: int64

In [16]:
((df.isnull().sum() / len(df) * 100) ).sort_values(ascending = False)


sales_commission_code    23.468457
category_name_1           0.028228
sku                       0.003593
status                    0.002737
Customer Since            0.002053
Customer ID               0.002053
price                     0.000171
item_id                   0.000171
created_at                0.000171
increment_id              0.000171
discount_amount           0.000171
grand_total               0.000171
qty_ordered               0.000171
Working Date              0.000171
payment_method            0.000171
BI Status                 0.000171
 MV                       0.000171
Month                     0.000171
Year                      0.000171
M-Y                       0.000171
FY                        0.000171
dtype: float64

- So `sales_commission_code` column have highest null values and this column is not important as well and the reason is that this column have no corelation with any other coolumn so lets drop it

In [17]:
df = df.drop('sales_commission_code', axis=1)